plugins/modules/cloud/misc/proxmox_snap.py

```diff
@@ -0,0 +1,289 @@
+
+
+VZ_TYPE = None
+
+
+def get_vmid(proxmox, hostname):
+    return [vm['vmid'] for vm in proxmox.cluster.resources.get(type='vm') if 'name' in vm and vm['name'] == hostname]
+
+
+def get_instance(proxmox, vmid):
+    return [vm for vm in proxmox.cluster.resources.get(type='vm') if int(vm['vmid']) == int(vmid)]
+
+
+def snapshot_create(module, proxmox, vm, vmid, timeout, snapname, description, vmstate):
+    if VZ_TYPE == 'lxc':
+        taskid = getattr(proxmox.nodes(vm[0]['node']), VZ_TYPE)(vmid).snapshot.post(snapname=snapname, description=description)
+    else:
+        taskid = getattr(proxmox.nodes(vm[0]['node']), VZ_TYPE)(vmid).snapshot.post(snapname=snapname, description=description, vmstate=int(vmstate))
+    while timeout:
+        if (proxmox.nodes(vm[0]['node']).tasks(taskid).status.get()['status'] == 'stopped' and
+                proxmox.nodes(vm[0]['node']).tasks(taskid).status.get()['exitstatus'] == 'OK'):
+            return True
+        timeout -= 1
+        if timeout == 0:
+            module.fail_json(msg='Reached timeout while waiting for creating VM snapshot. Last line in task before timeout: %s' %
+                                 proxmox.nodes(vm[0]['node']).tasks(taskid).log.get()[:1])
+
+        time.sleep(1)
+    return False
+
+
+def snapshot_remove(module, proxmox, vm, vmid, timeout, snapname, force):
+    taskid = getattr(proxmox.nodes(vm[0]['node']), VZ_TYPE)(vmid).snapshot.delete(snapname, force=int(force))
+    while timeout:
+        if (proxmox.nodes(vm[0]['node']).tasks(taskid).status.get()['status'] == 'stopped' and
+                proxmox.nodes(vm[0]['node']).tasks(taskid).status.get()['exitstatus'] == 'OK'):
+            return True
+        timeout -= 1
+        if timeout == 0:
+            module.fail_json(msg='Reached timeout while waiting for removing VM snapshot. Last line in task before timeout: %s' %
+                                 proxmox.nodes(vm[0]['node']).tasks(taskid).log.get()[:1])
+
+        time.sleep(1)
+    return False
+
+
+def setup_api(api_host, api_user, api_password, validate_certs):
+    api = ProxmoxAPI(api_host, user=api_user, password=api_password, verify_ssl=validate_certs)
+    return api
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            api_host=dict(required=True),
+            api_user=dict(required=True),
+            api_password=dict(no_log=True),
+            vmid=dict(required=False),
+            validate_certs=dict(type='bool', default='no'),
+            hostname=dict(),
+            timeout=dict(type='int', default=30),
+            state=dict(default='present', choices=['present', 'absent']),
+            description=dict(type='str'),
+            snapname=dict(type='str', default='ansible_snap'),
+            force=dict(type='bool', default='no'),
+            vmstate=dict(type='bool', default='no'),
+        )
+    )
+
+    if not HAS_PROXMOXER:
+        module.fail_json(msg='proxmoxer required for this module')
```

Comment:
```suggestion
        module.fail_json(msg='proxmoxer required for this module')
```
use `missed_required_lib` here, grep through other stuff

Reply:
Done!